h2o-py/h2o/utils/threading.py

```diff
@@ -0,0 +1,47 @@
+from __future__ import absolute_import
+
+import contextlib
+import threading
+
+_THREAD_CONTEXT = '_h2o_context_'
+
+
+@contextlib.contextmanager
+def thread_context(**kwargs):
+    """
+    Attach some key-value pairs to the current thread for the execution lifetime of the context block.
+    Attached keys can then be retrieved with the function ``thread_env``.
+    
+    :param kwargs: key-value pairs to be attached to the thread.
+    """
+    ct = threading.current_thread()
+    if not hasattr(ct, _THREAD_CONTEXT):
+        setattr(ct, _THREAD_CONTEXT, {})
+    context = getattr(ct, _THREAD_CONTEXT)
+    to_restore = {}
+    try:
+        for k, v in kwargs.items():
+            if k in context:
+                to_restore[k] = context[k]
+            context[k] = v
+        yield
+    finally:
+        for k, v in kwargs.items():
+            if k in to_restore:
+                context[k] = to_restore[k]
+            else:
+                del context[k]
```

Comment:
I'm not sure if this would be a real problem but this wouldn't prevent race-conditions when using `asyncio` but since Wave uses asyncio I'd rather be careful here and use [contextvars](https://docs.python.org/3/library/contextvars.html) when available. I think it could be an issue with IPython/Jupyter as it uses some asyncio eventloop as well but it's just a guess.

I used this artificial example to make sure it can cause race-conditions:
```python
import asyncio
import contextlib
import threading

_THREAD_CONTEXT = '_h2o_context_'


@contextlib.contextmanager
def thread_context(**kwargs):
    """
    Attach some key-value pairs to the current thread for the execution lifetime of the context block.
    Attached keys can then be retrieved with the function ``thread_env``.

    :param kwargs: key-value pairs to be attached to the thread.
    """
    ct = threading.current_thread()
    if not hasattr(ct, _THREAD_CONTEXT):
        setattr(ct, _THREAD_CONTEXT, {})
    context = getattr(ct, _THREAD_CONTEXT)
    to_restore = {}
    try:
        for k, v in kwargs.items():
            if k in context:
                to_restore[k] = context[k]
            context[k] = v
        yield
    finally:
        for k, v in kwargs.items():
            if k in to_restore:
                context[k] = to_restore[k]
            else:
                del context[k]


async def async_fun(sleep_duration):
    get_ctx = lambda : threading.current_thread()._h2o_context_ if hasattr(threading.current_thread(), _THREAD_CONTEXT) else None
    print(f"Before With: {get_ctx()}; sleep_duration: {sleep_duration}")
    with thread_context(sleep_duration=sleep_duration):
        print(f"Before Sleep: {get_ctx()}; sleep_duration: {sleep_duration}")
        await asyncio.sleep(sleep_duration)
        print(f"After Sleep: {get_ctx()}; sleep_duration: {sleep_duration}")
    print(f"After With: {get_ctx()}; sleep_duration: {sleep_duration}")


# await asyncio.gather(async_fun(1), async_fun(2)) # for jupyter/ipython as it has its own event loop

async def main():
    await asyncio.gather(async_fun(1), async_fun(2))

asyncio.run(main())
```

Reply:
Thanks for noticing this. I don't think it's a major concern in the context of this feature (concurrent prints in the event loop probably produces even worse results…) but as we don't know how exactly how the rendering can be done in Wave, it's worth handling this correctly and I'll try to use `contextvars` in the `thread_context` implementation when available (Py 3.7+).